src/ArrayHelper.php

```diff
@@ -428,20 +432,45 @@ public static function setValueByPath(array &$array, $path, $value, string $deli
 
     /**
      * @param array|float|int|string $path The path of where do you want to write a value to `$array`.
-     * The path can be described by a string when each key should be separated by delimiter.
+     * The path can be described by a string when each key should be separated by delimiter. If a path item contains
+     * delimiter, it can be escaped with "\" (backslash) or a custom delimiter can be used.
      * You can also describe the path as an array of keys.
-     * @param string $delimiter A separator, used to parse string $key for embedded object property retrieving. Defaults
+     * @param string $delimiter A separator, used to parse string key for embedded object property retrieving. Defaults
      * to "." (dot).
+     * @param bool $escapeDelimiter Whether to escape delimiter in the items of final array (in case of using string as
+     * an input) When `true`, "\" (backslashes) are removed - "\." will become "." for "." delimiter. Defaults to
+     * `false meaning backslashes will persist.
      *
      * @psalm-param ArrayPath $path
      *
      * @return array|float|int|string
      * @psalm-return ArrayKey
      */
-    private static function parsePath($path, string $delimiter)
+    public static function parsePath($path, string $delimiter = '.', bool $escapeDelimiter = false)
     {
         if (is_string($path)) {
-            return explode($delimiter, $path);
+            if (strlen($delimiter) !== 1) {
+                throw new InvalidArgumentException('Only 1 character is allowed for delimiter.');
+            }
+
+            if (($path[0] ?? '') === $delimiter) {
+                throw new InvalidArgumentException('Delimiter can\'t be at the very beginning.');
+            }
+
+            if (substr($path, -1) === $delimiter && substr($path, -2) !== '\\' . $delimiter) {
```

Comment:
In case `asdfasdfd\\.` string is end by delimiter, but don't pass this condition.

Reply:
We are only looking for non-escaped delimiter here. In your example it's escaped meaning it's a part of the string.